Improvements for handling default locale

vadz opened the pull request:
The changes here fix Mac-specific problems discussed in #23114, i.e.

1. Querying current locale before setting it now returns "C" there, just as everywhere else.
2. `wxLocale::Init(wxLANGUAGE_DEFAULT)` doesn't fail any longer with mixed locales.
3. While `GetSystemLocale()` still returns `wxLANGUAGE_UNKNOWN` for such locales, there is a new `GetSystemLocaleId()` function which allows to get e.g. "en-FR" string for them.

NB: this PR also incorporates #23119 and #23147.

cc @utelle @csomor (Stefan, I've removed Mac-specific code you added back in 93254327a1 (fixing strange bug under OS X (null strings from within CoreFoundation), 2007-12-19), but I'm pretty sure it's the right thing to do)

vadz committed:
Make comment in wxLocale::GetLanguageInfo() even more detailed

Explain that we need to call our (i.e. wxLocale) GetSystemLanguage()
here instead of letting wxUILocale::GetLanguageInfo() call its own (i.e.
wxUILocale) GetSystemLanguage() which does _not_ do the same thing.

No real changes, just try to explain this confusing code better.

vadz committed:
Add a pseudo test to show system locale and language

This can be useful to compare the results of calling
wxUILocale::GetSystemLanguage() and the function with the same name in
wxLocale (which actually corresponds to wxUILocale::GetSystemLocale()).

vadz committed:
Recommend using wxUILocale::GetSystemLanguage() in wxLocale docs

The wxUILocale function really returns the default language unlike the
one in wxLocale which must kepe using the default locale for
compatibility, making its name confusing.

vadz committed:
Handle wxLANGUAGE_DEFAULT specially in wxLocale::IsAvailable()

Check if using Init(wxLANGUAGE_DEFAULT) is going to succeed by actually
calling wxSetLocale(LC_ALL, "") in this special case, as we can't if
it's going to work otherwise if we don't recognize the language.

vadz committed:
Don't assume that wxLocale::Init(wxLANGUAGE_DEFAULT) succeeds

At least under macOS it fails when the system locale is something like
en_DE, for example, as setlocale() doesn't support such locales.

Use wxUILocale::GetSystemLanguage() instead of wxLANGUAGE_DEFAULT.

utelle committed:
Don't ignore LC_XXX in wxUILocaleImpl::GetPreferredUILanguages()

Previously, if LANGUAGE was set, the other locale-related environment
variables (LC_ALL, LC_MESSAGES, LANG) were simply ignored, which seems
wrong as the user running the program after explicitly setting LC_ALL
expects it to use this locale and not en_US from the default value of
LANGUAGE on a typical Linux system using English.

So handle LANGUAGE as a source of supplementary information, but still
honour all the other environment variables. This goes against an
explicit GNU gettext manual recommendation, but this recommendation
seems to only make sense in the usual case, when the first element of
LANGUAGE is the same as LC_ALL/LANG value, but not when they differ.

See #23146.

vadz committed:
Fix wxUILocale::FindLanguageInfo() to work for mixed locales

When looking for the language information, we must recognize the
language independently of the region it is followed by, so en_FR is
still English and fr_DE is still French, even if the full locale is
unknown, but this wasn't the case before.

Fix this by comparing the language part of wxLanguageInfo with just the
language of wxLocaleIdent we're trying to match, instead of comparing it
with its full BCP47 tag, which is never going to match.

vadz committed:
Return wxLANGUAGE_UNKNOWN from GetSystemLocale() in mixed locales

After the changes of the previous commit, GetSystemLocale() started
returning the language, found as due to the fallback logic implemented
in FindLanguageInfo(), for mixed locales, e.g. it would return English
for en_FR which is wrong as the actual locale is rather the French one,
using decimal comma and not period in this case.

Make it always return wxLANGUAGE_UNKNOWN for the mixed locales for now.

vadz committed:
Add wxUILocale::GetSystemLocaleId()

This function replaces the existing GetSystemLocale() as it can
represent the locales that don't have any corresponding wxLanguage
values and for which GetSystemLocale() has no choice but to return
wxLANGUAGE_UNKNOWN.

vadz committed:
Create valid (if empty) "C" locale under macOS

Use [NSLocale systemLocale] instead of trying to create a locale with
the name "C" which just fails and doesn't create any NSLocale at all.

If nothing else, this makes the behaviour of wxUILocale::GetCurrent()
consistent on all platforms as its GetLocaleId() now returns "C"
everywhere.

vadz committed:
Remove outdated special case for Mac from wxSetlocale()

This function must be used for setting the default CRT locale only and
we can't request arbitrary locales support from it under Mac as it
doesn't support any kind of mixed locales, such as "en-FR".

We do support them in wxUILocale, and calling wxUILocale::UseDefault()
from wxLocale::Init(wxLANGUAGE_DEFAULT) actually worked, but then this
function failed because wxSetlocale(LC_ALL, "") did not work for such
locales.

Fix this simply by removing Mac-specific code from this function.

This makes wxLocale::Init(wxLANGUAGE_DEFAULT) work as well as it ever
can under Mac.

utelle commented:
Although I tested only under Windows and Linux up to now, here comes a first feedback. Mac test will follow.

I tested with the _internat_ sample.

- Windows (system language German)
  - With language and locale `de-DE` all results as expected.
  - With language `de-DE` and regional formatting `fr-FR` German translations are used, except for the calendar control where French month and day names are used (but since the system was set to French regional formatting this was to be expected).
  - Testing for availability with `en_DE` results in "locale _en-DE / English (Germany)_ available" - and that's correct.
  - Testing for availability with `fr_DE` results in "locale _fr-DE_ is not available" - and that's correct.

- Linux (system language German)
  - `LC_ALL` **not** set: locale `de_DE.utf-8` is used for UI and C locale, all results as expected (except for a blank field under the heading **Date**)
  - Testing for availability with `en_DE` results in "locale _en_DE_ is available / Identifier _en_GB.utf-8 / English (United Kingdom)_" - and that is somewhat surprising. Why is the locale available, although it can't be set actually? Why is the identifier _en_GB_, although the identifier _en_DE_ is contained in the wx language database?
  - Testing for availability with `fr_DE` results in "locale _fr_DE_ is available / Identifier _fr_FR.utf-8 / French (France)_"" - again that is somewhat surprising. Why is the locale available, although it can't be set actually? Why is the identifier _fr_FR_?

  - `LC_ALL=de_DE`: locale `de_DE` is used for UI and C locale, all results as expected (except for a blank field under the heading **Date**)
  - `LC_ALL=en_DE`:
    - Setting `LC_ALL` like that results in a shell warning _setlocale: LC_ALL: can't change the locale settings (en_DE): File or directory not found
    - The program asks "Would you like to use the program in English (Germany)?" That is expected, because the locale `en_DE` is defined in the wx language database.
    - locale `C` is used for UI and C locale, all results in English as expected (except for a blank field under the heading **Date**)
  - `LC_ALL=fr_DE`:
    - Setting `LC_ALL` like that results in a shell warning _setlocale: LC_ALL: can't change the locale settings (fr_DE): File or directory not found
    - The program asks "Would you like to use the program in English (United States)?" That is expected, because the locale `fr_DE` is **not** defined in the wx language database.
    - locale `C` is used for UI and C locale, all results in English as expected (except for a blank field under the heading **Date**)

  IMHO the locales like `en_DE` and `fr_DE` should be reported as **not available** under Linux, because they can't be set via `setlocale`. If they are accepted, then most likely the expected behaviour would be to use the language part for the UI language and the region part for the regional formatting. However, this doesn't happen, because the actual locale will be _C_.

I will have to look at the code changes for Linux to understand what is going on.

vadz commented:
Thanks for testing! I think the discussion of support for "mixed" locales under Linux should be restricted to #23217. And there is also the more specific #23218 for the "locale is unexpectedly available" part of the problem.

Other than that it looks like you haven't found any problems, right?

Thanks again!

utelle commented:
Now I tested also under Mac OS Ventura. Again, I tested with the _internat_ sample.

- system settings `German (Germany)`
  - locale `de_DE` is used for UI, locale `C/UTF-8/C/C/C/C` is used for C locale.
  -  _Window_ menu and all subentries **not** translated (due to commit [f7d0742](https://github.com/wxWidgets/wxWidgets/commit/0a1b0fbe0142128447151c05a13103ed9f1d4fce) - message catalog **wxstd.pot** needs to be regenerated and **all** language catalogs need to be adjusted),
  - the calendar control uses **English** month and day names, numbers correctly use comma as decimal separator,
  -  the date field under the heading **Date** is blank instead of showing the current date,
  -  the locale name in English shows _Deutsch (Deutschland)_ instead of _German (Germany)_ - the required changes to `src/osx/core/uilocale.mm` were not yet applied.
  - Testing for availability with `en_DE` results in "locale _en_DE_ is available / Identifier _en_DE / English (Germany)_"
  - Testing for availability with `fr_DE` results in "locale _fr_DE_ is **not** available_"

- system settings `German (France)`
  -  locale `de_FR` is used for UI, locale `C/UTF-8/C/C/C/C` is used for C locale.
  - Testing for availability with `de_FR` results in "locale _de_FR_ is **not** available_" - although it corresponds to the system settings,
  - for locale name and calendar control same behaviour as above

- system settings `French (Germany)`
  -  locale `fr_DE` is used for UI, locale `C/UTF-8/C/C/C/C` is used for C locale.
  - Testing for availability with `fr_DE` results in "locale _fr_DE_ is **not** available_" - although it corresponds to the system settings,
  - for locale name and calendar control same behaviour as above

- system settings `English (Germany)`
  -  locale `en_DE` is used for UI, locale `C/UTF-8/C/C/C/C` is used for C locale.
  - Testing for availability with `en_DE` results in "locale _en_DE_ is available / Identifier _en_DE / English (Germany)_"
  - for locale name and calendar control same behaviour as above
  -  the date field under the heading **Date** is correctly shown formatted as a German date,

I find it interesting that the test for availability fails for `de_FR` and `fr_DE`, although the default locale was reported with exactly these values for the corresponding test cases. One would expect that a locale that is reported as the default locale would also be reported as available.

utelle commented:
> I think the discussion of support for "mixed" locales under Linux should be restricted to #23217. And there is also the more specific #23218 for the "locale is unexpectedly available" part of the problem.

IMHO under Linux the problem is that it is not detected when `setlocale` fails - the locale is reported as available nevertheless.

> Other than that it looks like you haven't found any problems, right?

There are a few minor issues:

- Linux: locales are reported as available, even if `setlocale` fails for them
- Linux and MacOS: the date field under the heading **Date** is blank quite often. If the language is English the current date is usually shown correctly.
- MacOS: method `GetLocalizedNames` not yet adjusted,
- MacOS: Window menu entries not translated
- MacOS: calendar control always uses English month and day names
- MacOS: some mixed locales reported as not available, even if the default locale is the same
- MacOS: all wxstd message catalogs need to be regenerated/adjusted

vadz committed:
Fail if environment variables define unknown locale under Unix

Don't always "succeed" in wxUILocale::UseDefault() and, consequently, in
wxLocale::Init(wxLANGUAGE_DEFAULT), under Unix systems, even if the
locale couldn't actually be set, as it can happen if the environment
variables contain a locale which is not supported on the current system,
e.g. a "mixed" locale such as "en_FR", or even a completely invalid
string such as "bloordyblop", which still used to succeed.

For now only fix it for reasonably modern systems with locale_t support,
it could be done even for the ancient ones without it later too if
anybody still cares about them.

Closes #23218.

vadz committed:
Fix GetSystemLocale() return value for "C" locale

This was broken by the recent changes, so fix it again: we do still want
to return wxLANGUAGE_ENGLISH_US for this particular locale.

vadz commented:
Concerning this part:

> Linux: locales are reported as available, even if `setlocale` fails for them

This is due to the code in `TryCreateMatchingLocale()` which explicitly looks for any locale with the same language, even if the region is different. I admit I have no idea why it was done, in spite of doing this myself in 3f8fd90c31 (Support creating Unix wxUILocale for languages without region, 2021-09-01) -- but the actual code doesn't do what the commit says, as it also creates locales even if we do have a different region.

Thinking more about this, I believe the idea here was to allow choosing a different region using the same language, i.e. if `fr_BE` is not supported, fall back on `fr_FR` if it is. This is probably still a good idea (although I'm not even 100% sure about this), but using `en_US` as fallback for `en_FR` almost surely is not, right?

utelle commented:
> Concerning this part:
> 
> > Linux: locales are reported as available, even if `setlocale` fails for them
> 
> This is due to the code in `TryCreateMatchingLocale()` which explicitly looks for any locale with the same language, even if the region is different. I admit I have no idea why it was done, in spite of doing this myself in [3f8fd90](https://github.com/wxWidgets/wxWidgets/commit/3f8fd90c31e290c8fdd6fbf5370f03ce5b1feea7) (Support creating Unix wxUILocale for languages without region, 2021-09-01) -- but the actual code doesn't do what the commit says, as it also creates locales even if we do have a different region.

IMHO a failing call to `setlocale` is a good indicator that a locale shoud be reported as "not available". 

AFAICR the approach in pre-wxUILocale times was similar. However, I find this approach questionable. I didn't change it to not break wxLocale compatibility, but regarding wxUILocale I would rather drop this fallback "strategy".

I'm not sure what the actual expectations of a user are, when he/she uses a "mixed" locale of which it is known that it is not supported by the OS.

The reason for setting a locale like `fr_US` or `de_US` is usually that the user is located in the given region, but doesn't speak well the region's main language and therefore prefers to use his native language for the UI, but a regional formatting, so that other people do not misinterpret dates and numbers.

> Thinking more about this, I believe the idea here was to allow choosing a different region using the same language, i.e. if `fr_BE` is not supported, fall back on `fr_FR` if it is. This is probably still a good idea (although I'm not even 100% sure about this), but using `en_US` as fallback for `en_FR` almost surely is not, right?

Well, I'm not really convinced that falling back on a locale with the same language, but a different region is such a good idea. Although it seems to be rather plausible to use `fr_FR` instead of `fr_BE` (France and Belgium are neighboring countries after all), it is absolutely not plausible to fall back from `en_DE` to `en_GB` - date and number formatting are very different in Germany and the United Kingdom.

Under Windows many language/region combinations are predefined, but not all. Nevertheless, you can set the locale to almost any combination programmatically, even though the settings dialog doesn't offer it. Windows "constructs" internally a locale that tries to match the language and the region as close as possible. For example, for `fr_US` it would use the French language, but date and number formatting would follow the US rules. This is far better than falling back to a different locale replacing the region in a somewhat obscure manner. However, our code under Windows reports "constructed" locales as "not available". And IMHO that is a good decision.

I think the most logical approach is to allow for date and number formatting (and similar locale dependent things) to set a locale that is actually supported by the OS, and to let the user choose a different language for the UI. This should cover most use cases.

vadz committed:
Remove fallback on locale using different region under Unix

Historically, we used to fall back on any locale using the same
language, if the exactly requested locale was unavailable. This doesn't
seem such a great idea, and results in unexpected behaviour, such as
returning true from wxUILocale::IsSupported() for locales such as
"en-FR" that are not really supported at all.

Remove this fallback and just return false if the locale with the
specified region is not supported.

Still choose any supported region if the region is omitted entirely,
however.

vadz commented:
The trouble is that I'd like to apply these fixes to 3.2 too and removing this fallback would definitely be a compatibility breakage, so I don't want to do it there. But I agree that we should stop doing this in master, so I've added a commit doing this.

vadz commented:
Finally, concerning the locales availability under Mac: it turns out `en-FR` is supported by Cocoa, i.e. it's one of the locales returned by `[NSLocale availableLocaleIdentifiers]`, while neither `fr-DE` nor `de-FR` are in [this list](https://github.com/wxWidgets/wxWidgets/files/10670079/macos-12-locales.txt).

I don't know if it's worth handling the current locale specially, maybe it is, but for now I won't do it.

utelle commented:
> Finally, concerning the locales availability under Mac: it turns out `en-FR` is supported by Cocoa, i.e. it's one of the locales returned by `[NSLocale availableLocaleIdentifiers]`, while neither `fr-DE` nor `de-FR` are in [this list](https://github.com/wxWidgets/wxWidgets/files/10670079/macos-12-locales.txt).

Question: Should we update our language database to include the locales from the above list that are not yet included?

> I don't know if it's worth handling the current locale specially, maybe it is, but for now I won't do it.

Since on MacOS there is no way to set (`Use()`) a locale other than the default locale, the user can always access the current locale to retrieve the information. So, for now it is ok as it is.

vadz commented:
> Question: Should we update our language database to include the locales from the above list that are not yet included?

I don't know how does macOS determine which locales to support, so probably not -- as I don't see how are we going to maintain it in the future. But if anybody has a source for this data, then perhaps we should...

vadz committed:
Don't use LC_MESSAGES for determining locale on non-glibc systems

LC_MESSAGES should only be used for getting the translations language,
so using GetLocaleFromEnvironment(), which uses its value if LC_ALL was
not set, in wxUILocaleImplUnix::InitLocaleNameAndCodeset() was wrong for
the cases when we did call it from there, i.e. when not using glibc and
so _NL_LOCALE_NAME is not defined.

Correct this by replacing GetLocaleFromEnvironment() querying all
environment variables at once with GetLocaleFromEnvVar() getting the
value from just the one environment variable specified by caller and
not calling it for LC_MESSAGES from InitLocaleNameAndCodeset().

See #23217.

vadz commented:
I'm going to merge this without waiting for more reviews/testing because I want to backport it to 3.2 and for this it should be merged into master first, but any feedback would be still very welcome even later!
